### PR TITLE
PARQUET-502: Ensure that a scanner reads the entire column regardless…

### PR DESCRIPTION
**Pull request by asandryh**
… of the batch size.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -279,4 +279,5 @@ add_custom_target(clean-all
 # installation
 
 install(TARGETS parquet
-  LIBRARY DESTINATION lib)
+  LIBRARY DESTINATION lib
+  ARCHIVE DESTINATION lib)
```
Comment (asandryh) [inline CMakeLists.txt]: This is to enable static builds. Not relevant to parquet-502, but useful in general, and not worth opening a separate jira.

File: src/parquet/column/scanner.h [hunk 2]
```diff
@@ -108,6 +108,7 @@ class TypedScanner : public Scanner {
 
       // TODO: repetition levels
 
+      value_offset_ = 0;
```
Comment (asandryh) [inline src/parquet/column/scanner.h]: That's the whole fix. :)

File: src/parquet/reader.cc [hunk 3]
```diff
@@ -288,17 +288,18 @@ void ParquetFileReader::DebugPrint(std::ostream& stream, bool print_values) {
     }
     stream << "\n";
 
-    bool hasRow;
-    do {
-      hasRow = false;
+    for (size_t r = group_reader->column_metadata(0)->num_values; r >0; r--) {
```
Comment (asandryh) [inline src/parquet/reader.cc]: This is a cosmetic change, to prevent DebugPrint from printing an empty row at the end of the output.

**Comment by asandryh** [comment]
Passing CI build: https://travis-ci.org/asandryh/parquet-cpp/builds/107864253

File: src/parquet/reader-test.cc [hunk 4]
```diff
@@ -84,36 +84,56 @@ TEST_F(TestAllTypesPlain, TestBatchRead) {
...
+  int32_t val;
+  bool is_null;
+  for (size_t i = 0; i < 8; ++i) {
+    ASSERT_TRUE(scanner.HasNext());
+    ASSERT_TRUE(scanner.NextValue(&val, &is_null));
+    ASSERT_FALSE(is_null);
+  }
+  ASSERT_FALSE(scanner.HasNext());
+  ASSERT_FALSE(scanner.NextValue(&val, &is_null));
+}
```
Comment (wesm) [inline src/parquet/reader-test.cc]: It's a bummer that this only works with REQUIRED data and files with 8 values. I feel like we're digging the technical debt hole deeper by not taking this opportunity to write unit tests vs increasing our reliances on binary files. Thoughts?

File: src/parquet/reader-test.cc [hunk 5]
```diff
@@ -84,36 +84,56 @@ TEST_F(TestAllTypesPlain, TestBatchRead) {
   ASSERT_FALSE(col->HasNext());
 }
 
+void verifyScanner(Int32Scanner& scanner) {
```
Comment (wesm) [inline src/parquet/reader-test.cc]: camel case

File: src/parquet/reader.cc [hunk 6]
```diff
@@ -288,17 +288,18 @@ void ParquetFileReader::DebugPrint(std::ostream& stream, bool print_values) {
...
       for (int i = 0; i < num_columns; ++i) {
         if (scanners[i]->HasNext()) {
-          hasRow = true;
           scanners[i]->PrintNext(stream, 17);
+        } else {
+          std::string null_fmt = format_fwf<Type::BYTE_ARRAY>(17);
+          snprintf(buffer, bufsize, null_fmt.c_str(), "NULL");
+          stream << buffer;
```
Comment (wesm) [inline src/parquet/reader.cc]: When does this happen?

File: src/parquet/reader-test.cc (hunk 4 above)
Comment (wesm) [inline src/parquet/reader-test.cc]: My specific point is that we really need a unit test that reproduces this bug, and so the moment we nix these temporary tests we have to go back and retroactively write the unit tests we probably should have written the first time around. I only wrote these tests here to help me do the refactoring in PARQUET-433 (which would have severely blocked progress if not gotten through quickly), but not as a long term solution.

**Comment by majetideepak** [comment]
I am going spend the next couple of days adding some unit tests.

**Comment by wesm** [comment]
OK, maybe combine with this patch? I'm working on other things the next 2 days so let me know if I can assist with test fixtures etc.

File: src/parquet/reader-test.cc (hunk 4 above)
Comment (asandryh) [inline src/parquet/reader-test.cc]: I have started working on PARQUET-481 with the goal to remove the reliance on the files in /data. I'll submit a separate PR for that. I can take care of that first, and then rebase this PR. But until then we need to be aware of this bug in TypedScanner.

File: src/parquet/reader.cc (hunk 6 above)
Comment (asandryh) [inline src/parquet/reader.cc]: When some of the columns have NULLs at the end.

**Comment by asandryh** [comment]
@majetideepak Do you plan to add tests for FileReader and Scanner? I'm doing this as a part of PARQUET-481. Let's make sure we don't do the same work twice.

**Comment by majetideepak** [comment]
@asandryh Please go ahead and implement the tests for FileReader and
Scanner.

On Mon, Feb 8, 2016 at 8:24 PM, asandryh <notifications@github.com> wrote:

> @majetideepak <https://github.com/majetideepak> Do you plan to add tests
> for FileReader and Scanner? I'm doing this as a part of PARQUET-481. Let's
> make sure we don't do the same work twice.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/parquet-cpp/pull/43#issuecomment-181658740>.
>



-- 
regards,
Deepak Majeti

**Comment by wesm** [comment]
@asandryh @majetideepak can you give a status update? This isn't strictly on the immediate critical path but will need to be addressed as part of more exhaustive Scanner test coverage https://issues.apache.org/jira/browse/PARQUET-526. This also overlaps with https://issues.apache.org/jira/browse/PARQUET-499

thank you

**Comment by asandryh** [comment]
I've created a new PR for PARQUET-526: https://github.com/apache/parquet-cpp/pull/50. It also contains the fix for PARQUET-502.

**Comment by majetideepak** [comment]
@asandryh can you please rebase your PRs. Have you started working on PARQUET-481 ? If not, we could work on it together.

**Comment by wesm** [comment]
@asandryh @majetideepak I'm pretty tied up with the Spark Summit East conference this week so I will not have much time for Parquet until Friday at earliest, but we need to prioritize completing the PLAIN encoding and decoding routines and tests before continuing much further down this testing path. For example, the bug that is fixed here is difficult to test because the encoding tools for  `FIXED_LEN_BYTE_ARRAY` are incomplete (or at least untested).

**Comment by wesm** [comment]
I updated the planning document to group interrelated JIRAs

https://docs.google.com/document/d/1WyquzupLc3UkErO2OhqLJNQ9a84Cccc8LVUSuLQz39o/edit#

I recommend completing PARQUET-499 before working more on this patch or #50

**Comment by majetideepak** [comment]
This is a  good plan.

**Pull request closed without merging**